core/src/main/scala/org/apache/spark/storage/BlockManagerMessages.scala

```diff
@@ -108,4 +107,6 @@ private[spark] object BlockManagerMessages {
     extends ToBlockManagerMaster
 
   case class BlockManagerHeartbeat(blockManagerId: BlockManagerId) extends ToBlockManagerMaster
+
+  case class HasCachedBlocks(executorId: String) extends ToBlockManagerMaster
```

Comment:
Ah sorry, it was a bug. Check `BlockManagerMaster.hasCachedBlocks` method - instead of calling the same method recursively, in the context.askWithRetry, it should have been sending this message.